src/System.IO.Pipes/src/System/IO/Pipes/PipeCompletionSource.cs

```diff
@@ -0,0 +1,181 @@
+                    _cancellationRegistration = _cancellationToken.Register(thisRef => ((PipeCompletionSource<TResult>)thisRef).Cancel(), this);
+
+                    // Grab the state for case if IO completed while we were setting the registration.
+                    state = Interlocked.Exchange(ref _state, NoResult);
+                }
+                else if (state != CompletedCallback)
+                {
+                    // IO already completed and we have grabbed result state.
+                    // Set NoResult to prevent invocation of CompleteCallback(result state) from AsyncCallback(...)
+                    state = Interlocked.Exchange(ref _state, NoResult);
+                }
+
+                // If we have the result state of completed IO call CompleteCallback(result).
+                // Otherwise IO not completed.
+                if (state == ResultSuccess || state == ResultError)
+                {
+                    CompleteCallback(state);
+                }
+            }
+        }
+
+        internal void ReleaseResources()
+        {
+            _cancellationRegistration.Dispose();
+
+            // NOTE: The cancellation must *NOT* be running at this point, or it may observe freed memory
+            // (this is why we disposed the registration above)
+            if (Overlapped != null)
+            {
+                _threadPoolBinding.FreeNativeOverlapped(Overlapped);
+                _overlapped = null;
+            }
+        }
+
+        internal abstract void SetCompletedSynchronously();
+
+        protected virtual void AsyncCallback(uint errorCode, uint numBytes)
+        {
+            ErrorCode = (int)errorCode;
+
+            int resultState = errorCode == 0 ? ResultSuccess : ResultError;
+
+            // Store the result so that other threads can observe it
+            // and if no other thread is registering cancellation, continue.
+            // Otherwise CompleteCallback(resultState) will be invoked by RegisterForCancellation().
+            if (Interlocked.Exchange(ref _state, resultState) == NoResult)
+            {
+                // Now try to prevent invocation of CompleteCallback(resultState) from RegisterForCancellation().
+                // Otherwise, thread responsible for registering cancellation stole the result and it will invoke CompleteCallback(resultState).
+                if (Interlocked.Exchange(ref _state, CompletedCallback) != NoResult)
+                {
+                    CompleteCallback(resultState);
+                }
+            }
+        }
+
+        protected abstract void HandleError();
+
+        private void Cancel()
+        {
+            SafeHandle handle = _threadPoolBinding.Handle;
+            NativeOverlapped* overlapped = Overlapped;
+
+            // If the handle is still valid, attempt to cancel the IO
+            if (!handle.IsInvalid)
+            {
+                if (!Interop.mincore.CancelIoEx(handle, overlapped))
```

Comment:
This could be simplified slightly to
```C#
if (!handle.IsInvalid && !Interop.mincore.CancelIoEx(handle, overlapped))
{
    ...
}
```
rather than having the nested ifs.  Either ways fine, though.